packages/editor/src/lib/editor/Editor.ts

```diff
@@ -534,25 +538,33 @@ export class Editor extends EventEmitter<TLEventMap> {
 				},
 				binding: {
 					beforeCreate: (binding) => {
-						const next = this.getBindingUtil(binding).onBeforeCreate?.({ binding })
-						if (next) return next
+						if (this.isBindingValid(binding)) {
+							const next = this.getBindingUtil(binding).onBeforeCreate?.({ binding })
+							if (next) return next
+						}
 						return binding
 					},
 					afterCreate: (binding) => {
 						invalidBindingTypes.add(binding.type)
-						this.getBindingUtil(binding).onAfterCreate?.({ binding })
+						if (this.ensureBindingValid(binding)) {
+							this.getBindingUtil(binding).onAfterCreate?.({ binding })
+						}
 					},
 					beforeChange: (bindingBefore, bindingAfter) => {
-						const updated = this.getBindingUtil(bindingAfter).onBeforeChange?.({
-							bindingBefore,
-							bindingAfter,
-						})
-						if (updated) return updated
+						if (this.ensureBindingValid(bindingAfter)) {
```

Comment:
I don't think this should run on `beforeChange` or `beforeCreate`, only `afterChange` and `afterCreate`. Otherwise it makes updates rely on the order in which updates are applied during e.g. applyDiff

    applyDiff({ `add shape a`, `add binding between shape a and b`, `add shape b`})

if those diff operations were applied in the order I wrote them, the binding's `onBeforeCreate` would not be invoked.

    applyDiff({ `change binding toId to shape b`, `add shape b` })

and here the binding would end up being deleted instead of updated.